crates/re_log_types/src/data_cell.rs

```diff
@@ -0,0 +1,445 @@
+    ///
+    /// Internally this is always stored as an erased arrow array to avoid bad surprises with
+    /// frequent boxing/unboxing down the line.
+    pub(crate) values: Box<dyn arrow2::array::Array>,
+}
+
+// TODO(cmc): We should be able to build a cell from non-reference types.
+// TODO(#1619): We shouldn't have to specify the component name separately, this should be
+// part of the metadata by using an extension.
+// TODO(#1696): Check that the array is indeed a leaf / component type when building a cell from an
+// arrow payload.
+impl DataCell {
+    /// Builds a new `DataCell` from a uniform iterable of native component values.
+    ///
+    /// Fails if the given iterable cannot be serialized to arrow, which should never happen when
+    /// using Rerun's builtin components.
+    #[inline]
+    pub fn try_from_native<'a, C: SerializableComponent>(
+        values: impl IntoIterator<Item = &'a C>,
+    ) -> DataCellResult<Self> {
+        use arrow2_convert::serialize::TryIntoArrow;
+        Ok(Self::from_arrow(
+            C::name(),
+            TryIntoArrow::try_into_arrow(values.into_iter())?,
+        ))
+    }
+
+    /// Builds a new `DataCell` from a uniform iterable of native component values.
+    ///
+    /// Fails if the given iterable cannot be serialized to arrow, which should never happen when
+    /// using Rerun's builtin components.
+    #[inline]
+    pub fn try_from_native_sparse<'a, C: SerializableComponent>(
+        values: impl IntoIterator<Item = &'a Option<C>>,
+    ) -> DataCellResult<Self> {
+        use arrow2_convert::serialize::TryIntoArrow;
+        Ok(Self::from_arrow(
+            C::name(),
+            TryIntoArrow::try_into_arrow(values.into_iter())?,
+        ))
+    }
+
+    /// Builds a new `DataCell` from a uniform iterable of native component values.
+    ///
+    /// Panics if the given iterable cannot be serialized to arrow, which should never happen when
+    /// using Rerun's builtin components.
+    /// See [`Self::try_from_native`] for the fallible alternative.
+    #[inline]
+    pub fn from_native<'a, C: SerializableComponent>(
+        values: impl IntoIterator<Item = &'a C>,
+    ) -> Self {
+        Self::try_from_native(values).unwrap()
+    }
+
+    /// Builds a new `DataCell` from a uniform iterable of native component values.
+    ///
+    /// Panics if the given iterable cannot be serialized to arrow, which should never happen when
+    /// using Rerun's builtin components.
+    /// See [`Self::try_from_native`] for the fallible alternative.
+    #[inline]
+    pub fn from_native_sparse<'a, C: SerializableComponent>(
+        values: impl IntoIterator<Item = &'a Option<C>>,
+    ) -> Self {
+        Self::try_from_native_sparse(values).unwrap()
+    }
+
+    /// Builds a cell from an iterable of items that can be turned into a [`Component`].
+    ///
+    /// ⚠ This requires allocating due to quirks in `arrow2-convert`, prefer [`Self::from_native`]
```

Comment:
doesn't `from_native` also allocate? Do you mean this requires an _extra_ allocation?